### ISSUE TITLE
FrozenError: can't modify frozen Hash on /events when `Pubsubstub.use_persistent_connections = false`

### ISSUE THREAD

**Issue by ohsabry** [Shopify/shipit-engine#1081]
We recently attempted to running pubsubstub in the integrated polling mode, because after upgrading to rails 6 and shipit 0.31 I ran into some issues with redis and sessions with my pubsubstub config (will make a separate issue - though this integrated mode seems easier to manage for our needs). 

I see that the client is now polling this endpoint: `/events?channels[]=stack.$stack_id` - however, the server is responding with a 500 every time.

Here is the stack trace on the server:

I am running rails 6.0.3.1, shipit 0.31.0 and pubstub 0.2.0

```
FrozenError: can't modify frozen Hash
  from rack/etag.rb:35:in `call'
  from rack/conditional_get.rb:27:in `call'
  from rack/head.rb:12:in `call'
  from action_dispatch/http/content_security_policy.rb:18:in `call'
  from rack/session/abstract/id.rb:266:in `context'
  from rack/session/abstract/id.rb:260:in `call'
  from action_dispatch/middleware/cookies.rb:648:in `call'
  from action_dispatch/middleware/callbacks.rb:27:in `block in call'
  from active_support/callbacks.rb:101:in `run_callbacks'
  from action_dispatch/middleware/callbacks.rb:26:in `call'
  from action_dispatch/middleware/actionable_exceptions.rb:17:in `call'
  from action_dispatch/middleware/debug_exceptions.rb:32:in `call'
  from action_dispatch/middleware/show_exceptions.rb:33:in `call'
  from rails/rack/logger.rb:37:in `call_app'
  from rails/rack/logger.rb:26:in `block in call'
  from active_support/tagged_logging.rb:80:in `block in tagged'
  from active_support/tagged_logging.rb:28:in `tagged'
  from active_support/tagged_logging.rb:80:in `tagged'
  from rails/rack/logger.rb:26:in `call'
  from action_dispatch/middleware/remote_ip.rb:81:in `call'
  from action_dispatch/middleware/request_id.rb:27:in `call'
  from rack/method_override.rb:24:in `call'
  from rack/runtime.rb:22:in `call'
  from active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
  from action_dispatch/middleware/executor.rb:14:in `call'
  from action_dispatch/middleware/static.rb:126:in `call'
  from rack/sendfile.rb:110:in `call'
  from action_dispatch/middleware/host_authorization.rb:76:in `call'
  from raven/integrations/rack.rb:51:in `call'
  from rails/engine.rb:527:in `call'
  from puma/configuration.rb:227:in `call'
  from puma/server.rb:706:in `handle_request'
  from puma/server.rb:476:in `process_client'
  from puma/server.rb:334:in `block in run'
  from puma/thread_pool.rb:135:in `block in spawn_thread'
```

Any ideas? We are running puma with one worker and multiple threads

**Comment by ohsabry** [comment]
Extreme hack, but downgrading to rack 2.0.1 fixed it (freeze was added in 2.0.2)

**Comment by casperisfine** [comment]
It's weird, line 35 is trying to set a header in the headers hash: https://github.com/rack/rack/blob/a5e80f01947954af76b14c1d1fdd8e79dd8337f3/lib/rack/etag.rb#L35

This hash being frozen is quite uncommon. 

However this doesn't seem related to Shipit at all. So I'll close.  I however encourage you to reproduce this in a small project and report the bug to rack.

**Comment by josacar** [comment]
Just booted up in development mode and I have the same when I updated to `0.31.0`:

```
::1 - - [26/Sep/2020:17:25:47 CEST] "GET /events?channels%5B%5D=stack.1 HTTP/1.1" 500 58037
http://localhost:3000/josacar/glare/production -> /events?channels%5B%5D=stack.1
can't modify frozen Hash: {"Content-Type"=>"text/event-stream", "Cache-Control"=>"no-cache", "X-Accel-Buffering"=>"no", "Connection"=>"keep-alive"} excluded from capture: DSN not set

FrozenError (can't modify frozen Hash: {"Content-Type"=>"text/event-stream", "Cache-Control"=>"no-cache", "X-Accel-Buffering"=>"no", "Connection"=>"keep-alive"}):

rack (2.2.3) lib/rack/etag.rb:35:in `call'
rack (2.2.3) lib/rack/conditional_get.rb:27:in `call'
rack (2.2.3) lib/rack/head.rb:12:in `call'
actionpack (6.0.3.3) lib/action_dispatch/http/content_security_policy.rb:18:in `call'
rack (2.2.3) lib/rack/session/abstract/id.rb:266:in `context'
rack (2.2.3) lib/rack/session/abstract/id.rb:260:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/cookies.rb:648:in `call'
activerecord (6.0.3.3) lib/active_record/migration.rb:567:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
activesupport (6.0.3.3) lib/active_support/callbacks.rb:101:in `run_callbacks'
actionpack (6.0.3.3) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/actionable_exceptions.rb:17:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/debug_exceptions.rb:32:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
lograge (0.11.2) lib/lograge/rails_ext/rack/logger.rb:15:in `call_app'
railties (6.0.3.3) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (6.0.3.3) lib/active_support/tagged_logging.rb:80:in `block in tagged'
activesupport (6.0.3.3) lib/active_support/tagged_logging.rb:28:in `tagged'
activesupport (6.0.3.3) lib/active_support/tagged_logging.rb:80:in `tagged'
railties (6.0.3.3) lib/rails/rack/logger.rb:26:in `call'
sprockets-rails (3.2.2) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/request_id.rb:27:in `call'
rack (2.2.3) lib/rack/method_override.rb:24:in `call'
rack (2.2.3) lib/rack/runtime.rb:22:in `call'
activesupport (6.0.3.3) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/static.rb:126:in `call'
rack (2.2.3) lib/rack/sendfile.rb:110:in `call'
actionpack (6.0.3.3) lib/action_dispatch/middleware/host_authorization.rb:82:in `call'
sentry-raven (2.13.0) lib/raven/integrations/rack.rb:51:in `call'
webpacker (4.3.0) lib/webpacker/dev_server_proxy.rb:23:in `perform_request'
rack-proxy (0.6.5) lib/rack/proxy.rb:57:in `call'
railties (6.0.3.3) lib/rails/engine.rb:527:in `call'
rack (2.2.3) lib/rack/handler/webrick.rb:95:in `service'
/Users/joseluis/.asdf/installs/ruby/2.7.1/lib/ruby/2.7.0/webrick/httpserver.rb:140:in `service'
/Users/joseluis/.asdf/installs/ruby/2.7.1/lib/ruby/2.7.0/webrick/httpserver.rb:96:in `run'
/Users/joseluis/.asdf/installs/ruby/2.7.1/lib/ruby/2.7.0/webrick/server.rb:307:in `block in start_thread'
::1 - - [26/Sep/2020:17:25:52 CEST] "GET /events?channels%5B%5D=stack.1 HTTP/1.1" 500 58037
http://localhost:3000/josacar/glare/production -> /events?channels%5B%5D=stack.1
``

**Comment by josacar** [comment]
Looks it's due to [this](https://github.com/gmalette/pubsubstub/blob/a445c4969f2a7a1fe0894320f1ecd1d7e362d611/lib/pubsubstub/stream_action.rb#L3-L8) freeze

**Comment by casperisfine** [comment]
It's kind of a bug in Rack::Etag, up to 2.2.3 it expects the headers hash to be mutable, it's fixed in https://github.com/rack/rack/pull/1641, but that isn't released yet.

**Comment by casperisfine** [comment]
Fixed in pubsubstub 0.2.1.

**Comment by josacar** [comment]
Thanks for taking a look, will update.
